include/utils/math/estimator/srukf.h

```diff
@@ -0,0 +1,616 @@
+        // multiple measurement updates per time update
+        Eigen::Matrix<double, ROWS, STATES + 2> sigmas_H;
+        Eigen::Matrix<double, STATES, STATES + 2> sigmas = m_pts.square_root_sigma_points(m_xhat, m_S);
+        for (int i = 0; i < m_pts.num_sigmas(); ++i) {
+            sigmas_H.template block<ROWS, 1>(0, i) = h(sigmas.template block<STATES, 1>(0, i), u);
+        }
+
+        // Pass the predicted measurement sigmas through the Unscented Transform
+        // to compute the mean predicted measurement and square-root innovation
+        // covariance.
+        //
+        // equations (23) (24) and (25)
+        auto [yhat, Sy] =
+          square_root_ut<ROWS, STATES>(sigmas_H, m_pts.Wm(), m_pts.Wc(), mean_func_Y, residual_func_Y, sqrt_R.template triangularView<Eigen::Lower>());
+
+        // Compute cross covariance of the predicted state and measurement sigma
+        // points given as:
+        //
+        //           n+1
+        //   P_{xy} = Σ Wᵢ⁽ᶜ⁾[𝒳ᵢ - x̂][𝒴ᵢ - ŷ⁻]ᵀ
+        //           i=0
+        //
+        // equation (26)
+        Eigen::Matrix<double, STATES, ROWS> Pxy;
+        Pxy.setZero();
+        for (int i = 0; i < m_pts.num_sigmas(); ++i) {
+            Pxy += m_pts.Wc(i) * (residual_func_X(m_sigmas_F.template block<STATES, 1>(0, i), m_xhat)) *
+                   (residual_func_Y(sigmas_H.template block<ROWS, 1>(0, i), yhat)).transpose();
+        }
+
+        // Compute the Kalman gain. We use Eigen's QR decomposition to solve. This
+        // is equivalent to MATLAB's \ operator, so we need to rearrange to use
+        // that.
+        //
+        //   K = (P_{xy} / S_{y}ᵀ) / S_{y}
+        //   K = (S_{y} \ P_{xy})ᵀ / S_{y}
+        //   K = (S_{y}ᵀ \ (S_{y} \ P_{xy}ᵀ))ᵀ
+        //
+        // equation (27)
+        Eigen::Matrix<double, STATES, ROWS> K = (Sy.transpose().fullPivHouseholderQr().solve(Sy.fullPivHouseholderQr().solve(Pxy.transpose()))).transpose();
+
+        // Compute the posterior state mean
+        //
+        //   x̂ = x̂⁻ + K(y − ŷ⁻)
+        //
+        // second part of equation (27)
+        m_xhat = add_func_X(m_xhat, K * residual_func_Y(y, yhat));
+
+        // Compute the intermediate matrix U for downdating
+        // the square-root covariance
+        //
+        // equation (28)
+        Eigen::Matrix<double, STATES, ROWS> U = K * Sy;
+
+        // Downdate the posterior square-root state covariance
+        //
+        // equation (29)
+        for (int i = 0; i < ROWS; i++) {
+            Eigen::internal::llt_inplace<double, Eigen::Lower>::rankUpdate(m_S, U.template block<STATES, 1>(0, i), -1);
+        }
+    }
+
+  private:
+    std::function<StateVector(const StateVector &, const InputVector &)> m_f;
+    std::function<OutputVector(const StateVector &, const InputVector &)> m_h;
+    std::function<StateVector(const Eigen::Matrix<double, STATES, STATES + 2> &, const Eigen::Vector<double, STATES + 2> &)> m_mean_func_X;
```

Comment:
This and the following line have a lot going on. Is there a handy name to give this type of functions. anything like
```c++
using XMeanFunc = std::function<...>;
using YMeanFunc = std::function<...>;
...
XResidualFunc m_mean_func_x;
YResidualFunc m_mean_func_y
```
Also, STATES+2 is the number of sigma points, right? Since you're writing STATES+2 everywhere it might be handy to have a name for that and make a (static constexpr) member like
```c++
class SRUKF{
     static constexpr int NUM_SIGMA_PTS = STATES+2;
...
    Eigen::Matrix<double, STATES, NUM_SIGMA_PTS> m_sigmas_F;

} 
```
and etc, i think itll make it a little clearer what those matrices represent.

Reply:
or if you're feeling really crazy have ScaledSphericalSimplexSigmaPoints know how to calculate that then have SRUKF say `NUM_SIGMA_PTS = ScaledSphericalSimplexSigmaPoints<...>::thing`